# Exercise 2: Deep learning

CSCS-ICS-DADSi Summer School: Accelerating Data Science with HPC
September 4 - 6, 2017
Swiss National Supercomputing Centre

Mainly inspired from:

http://pytorch.org/tutorials/beginner/deep_learning_60min_blitz.html

We will use the PyTorch deep learning framework (http://pytorch.org/) in this exercise.

PyTorch is a Python-based reincarnation of the lua-based Torch framework (http://torch.ch/), with features such as dynamic graphs and general-purpose automatic differentiation.

Please visit the PyTorch website for more information and tutorials.

## Part 1: Tensors

We start by importing `torch`.

In [12]:
import torch
from torch import Tensor

Tensors, or multi-dimensional arrays, are the basic unit in PyTorch. Let's create a 5x3 matrix as a 2-dimensional Tensor. This will default to a `FloatTensor`, a tensor of single-precision (32-bits) floating point numbers. The tensor will be initialized with random numbers.

In [13]:
x = Tensor(5,3)
print(x)


 2.4105e-27  4.5717e-41  9.2910e-38
 0.0000e+00  0.0000e+00  9.8091e-45
-2.8088e+29  4.5716e-41  0.0000e+00
 0.0000e+00  0.0000e+00  0.0000e+00
 0.0000e+00  0.0000e+00  0.0000e+00
[torch.FloatTensor of size 5x3]



### Slicing

It is common to have higher-dimensional tensors, to hold all aspects of your data together, such as a 3x200x200 tensor holding the R,G,B channels of an image of 200x200 pixels, or a 100x3x200x200 tensor to hold a batch of 100 such images.

We use slicing operations to view specific subtensors of such a tensor. Note that PyTorch uses zero-based indexing and row-major memory ordering.

In [14]:
# 100 images of 3 channels of size 200x200
x = Tensor(100,3,200,200)
print(x.size())

# Second channel of the first image
y = x[0,1]
print(y.size())

torch.Size([100, 3, 200, 200])
torch.Size([200, 200])


### Numpy conversion

You might have noticed that Tensors are very similar to Numpy's ndarrays. In fact you can easily convert between ndarrays and Tensors.

In [15]:
x_np = x.numpy()
print(x_np.shape)

x = torch.from_numpy(x_np)
print(x.size())

(100, 3, 200, 200)
torch.Size([100, 3, 200, 200])


### A note on GPU usage

PyTorch is frequently used with a CUDA backend on a GPU. One would convert between tensors and CUDA tensors by simply calling

`x.cuda()`

`x.cpu()`

which transfer data to and from a GPU memory.

## Part 2: Automatic differentiation

PyTorch implements reverse-mode automatic differentiation (AD). For using AD features, we need the `Variable` class from  `autograd`.

In [8]:
from torch.autograd import Variable

For automatically evaluating derivatives with AD, we need to wrap tensors within `Variable` instances.

In [16]:
x = Tensor(3) # A vector of length three
x = Variable(x) 
print(x)

Variable containing:
 2.4105e-27
 4.5717e-41
 8.7568e-38
[torch.FloatTensor of size 3]



With `Variable`s, obtaining a gradient of a scalar-valued function is straightforward. Let's use the regular Python language features to define a function.

In [48]:
def f(x):
    y = torch.log(x[0]) + torch.sin(x[1])
    return x[2] * torch.exp(y)

Let's evaluate the gradient of `f(x)` at `x = [2,3,4]`.

In [50]:
x = Variable(Tensor([2,3,4]), requires_grad=True)
y = f(x)
print(y)

y.backward()
print(x.grad)

Variable containing:
 9.2125
[torch.FloatTensor of size 1]

Variable containing:
 4.6063
-9.1203
 2.3031
[torch.FloatTensor of size 3]



Note that we get the function evaluated `f(2,3,4)` as `9.2125` and the gradient `f'(2,3,4)` as `[4.6063, -9.1203, 2.3031]`.

Also note the use of `requires_grad = True` for tagging the `Variable` `x` for being differentiated with respect to. This is for efficiency reasons, allowing PyTorch to ignore the bookkeeping for `Variable`s whose derivatives we don't need.

## Part 3: Neural networks

The previous two parts, namely defining and using `Tensor`s and `Variable`s and being able to obtain derivatives, give us everything needed for constructing and training a neural network.

### Defining a neural network

A very important point to note is that **neural networks are just functions, and there is nothing special about them.**

A neural network is a series of linear algebra operations interleaved with non-linear transformations. Let's create a single feed-forward layer.

In [59]:
def run_layer(weights, inputs):
    outputs = torch.mm(weights , inputs)
    outputs = torch.tanh(outputs)
    return outputs

W = Variable(Tensor(2, 4), requires_grad=True) # A 2x4 weight matrix
x = Variable(Tensor([[1],[2],[3],[4]])) # A column vector of length four

y = run_layer(W, x)
print(y)

Variable containing:
1.00000e-27 *
  2.4105
  0.0000
[torch.FloatTensor of size 2x1]



In [55]:
torch.sigmoid(Tensor([1]))


 0.7311
[torch.FloatTensor of size 1]

By taking the gradient of a neural network with respect to its trainable parameters (weights), we can optimize it with gradient descent.

### The loss function

### Optimization

### Training data